### ISSUE TITLE
UTC is returned in local time 

### ISSUE THREAD

**Issue by Ufuk94** [simolus3/drift#286]
First of all, thanks for this amazing package and the hard work you put into it! 

I noticed that DateTime objects in UTC are returned in local time. Storing the date in Iso8601 format instead of MillisecondsSinceEpoch should solve this.

**Comment by simolus3** [comment]
> Storing the date in Iso8601 format instead of MillisecondsSinceEpoch should solve this.

That would have been a good idea, but unfortunately it'd be breaking for existing users now :(

As a starting point, you could use a custom type converter.
```dart
class IsoDateTimeConverter extends TypeConverter<DateTime, String> {
  const IsoDateTimeConverter();
  @override
  DateTime mapToDart(String fromDb) {
    if (fromDb == null) {
      return null;
    } else {
      return DateTime.parse(fromDb);
    }
  }

  @override
  String mapToSql(DateTime value) {
    return value?.toIso8601String();
  }
}
```
In your tables, you can now do:
```dart
// replace this
DateTimeColumn foo => dateTime()();
// with this:
TextColumn foo => text().map(const IsoDateTimeConverter())();
```

Arguably, moor should be able to do this out of the box. I'll think about how that could be done without breaking existing databases.

**Comment by kuhnroyal** [comment]
Maybe add a new ColumnType and moor file mapping?

**Comment by NightOwlCoder** [comment]
The problem I see with this approach is that we lose `DateTime` `where` and `orderBy` operations.

**Comment by JCKodel** [comment]
+1 for this.

Suggestion: In MSSQL, there is a special DateTime type called DateTimeOffset (it will record the date with timezone info and all queries will work as intended, no matter if comparing different timezones).

Moor could have a special type called DateTimeOffset or IsoDateTime that always convert the date to UTC while saving (so queries through different timezones will have no issues) and it always returns UTC date (so the developer knows that he must call value.toLocal() before using the date in the UI).

**Comment by fgblomqvist** [comment]
Wouldn't it make more sense if Moor just added a "storeAsString" option to the existing DateTime column rather than having a new type?

Regardless, I think @JCKodel is right in that the best way to avoid timezone issues is just for Moor to always covert to UTC and for the user to convert to the correct timezone when fetching. Doesn't make any sense whatsoever to rely on what timezone the object was stored in imo.

**Comment by simolus3** [comment]
Moor only stores timestamps, and `toLocal` or `toUtc` don't change the the timestamp, only the interpretation. The getters to extract individual fields from sql dates also return their values assuming the date was in GMT.

One might be able to use a type converter from `DateTime` to `DateTime` that just calls `toUtc` in both directions - it's doesn't change existing data and allows using the existing comparisons for `DateTime`. When using something like `column.day.equals(someDateTime.day)` one would still have to make sure that `someDateTime.isUtc` since it matters there.

But all of this is just workarounds, moor should have an option to store time with proper timezone information.

**Comment by JCKodel** [comment]
This is what I did:

```dart
import 'package:flutter/foundation.dart';
import 'package:moor/moor.dart';

@immutable
class DateTimeConverter extends TypeConverter<DateTime, int> {
  @override
  DateTime mapToDart(int fromDb) {
    if (fromDb == null) {
      return null;
    }

    final str = fromDb.toString();
    final millisecond = int.parse(str.substring(14, 17));
    final second = int.parse(str.substring(12, 14));
    final minute = int.parse(str.substring(10, 12));
    final hour = int.parse(str.substring(8, 10));
    final day = int.parse(str.substring(6, 8));
    final month = int.parse(str.substring(4, 6));
    final year = int.parse(str.substring(0, 4));

    return DateTime.utc(year, month, day, hour, minute, second, millisecond).toLocal();
  }

  @override
  int mapToSql(DateTime value) {
    if (value == null) {
      return null;
    }

    final date = value.toUtc();

    return date.millisecond +
        (date.second * 1000) +
        (date.minute * 100000) +
        (date.hour * 10000000) +
        (date.day * 1000000000) +
        (date.month * 100000000000) +
        (date.year * 10000000000000);
  }
}
```

Using int to store the uncompressed date is 55 bits (both Flutter and SQLite are 64 bits - but DartJS is only 53). 

This will use a int (which is faster than string and easier to compare and order) and the date will be readable (for instance `2020-05-14T20:39:40.567-03:00` will be saved as `20200514233940567`).

The converter ensures that all data in database are UTC and all data in Dart are local.

It should be easy to do queries such as:
```sql
// Get all sales from 2020
SELECT * FROM Sales WHERE date BETWEEN 20200000000000000 AND 20210000000000000;

// Get all sales this month
SELECT * FROM Sales WHERE date BETWEEN 20200500000000000 AND 20200600000000000;
```

**Comment by FLovison** [comment]
Very good job, @simolus3 , I love this package!
My cent on this question: I agree with store datetime in utc, so if I want filter with local datetime, I'm using this extensions,
that use sqlite function _DATE_ (or _DATETIME_) with 3th parameter _**localtime**_

```
extension LocalDateTimeExpressions on Expression<DateTime> {
  Expression<String> get dateLocal {
    return FunctionCallExpression(
      "DATE", // or DATETIME for date + time
      [
        this,
        const Constant<String>("unixepoch"),
        const Constant<String>("localtime")
      ],
    );
  }
}

extension LocalCompareDateDB on GeneratedDateTimeColumn {
  Expression<bool> dateLocalEquals(DateTime value) {
    return this.dateLocal.equals(value.toIso8601String().substring(0, 10));
  }
}
```

So I can use this method into where SQL:

`
query.where(tableName.fieldDate.dateLocalEquals(DateTime.now()));`

**Comment by mahiznan** [comment]
> Very good job, @simolus3 , I love this package!
> My cent on this question: I agree with store datetime in utc, so if I want filter with local datetime, I'm using this extensions,
> that use sqlite function _DATE_ (or _DATETIME_) with 3th parameter _**localtime**_
> 
> ```dart
> extension LocalDateTimeExpressions on Expression<DateTime> {
>   Expression<String> get dateLocal {
>     return FunctionCallExpression(
>       "DATE", // or DATETIME for date + time
>       [
>         this,
>         const Constant<String>("unixepoch"),
>         const Constant<String>("localtime")
>       ],
>     );
>   }
> }
> 
> extension LocalCompareDateDB on GeneratedDateTimeColumn {
>   Expression<bool> dateLocalEquals(DateTime value) {
>     return this.dateLocal.equals(value.toIso8601String().substring(0, 10));
>   }
> }
> ```
> 
> So I can use this method into where SQL:
> 
> ```dart
> query.where(tableName.fieldDate.dateLocalEquals(DateTime.now()));
> ```

@FLovison Where to add this extension? Any sample code?

**Comment by FLovison** [comment]
Sample MoveItem table with moveDate column of DateTime type:

```dart

import 'package:moor/moor.dart';
import 'utils.dart';

@DataClassName("MoveItem")
class MoveItems extends Table {
  String get tableName => "moves";

  IntColumn get id => integer().autoIncrement()();
  TextColumn get description => text().nullable()();
  DateTimeColumn get moveDate => dateTime().nullable()();
}

// A simple select with dateEquals method

Stream<List<MoveItem>> moveWatch() {
  return select(moveItems).moveDate.dateEquals(DateTime.now());
}

```

**Comment by akashroshan135** [comment]
> 
> 
> Very good job, @simolus3 , I love this package!
> My cent on this question: I agree with store datetime in utc, so if I want filter with local datetime, I'm using this extensions,
> that use sqlite function _DATE_ (or _DATETIME_) with 3th parameter _**localtime**_
> 
> ```dart
> extension LocalDateTimeExpressions on Expression<DateTime> {
>   Expression<String> get dateLocal {
>     return FunctionCallExpression(
>       "DATE", // or DATETIME for date + time
>       [
>         this,
>         const Constant<String>("unixepoch"),
>         const Constant<String>("localtime")
>       ],
>     );
>   }
> }
> 
> extension LocalCompareDateDB on GeneratedDateTimeColumn {
>   Expression<bool> dateLocalEquals(DateTime value) {
>     return this.dateLocal.equals(value.toIso8601String().substring(0, 10));
>   }
> }
> ```
> 
> So I can use this method into where SQL:
> 
> ```dart
> query.where(tableName.fieldDate.dateLocalEquals(DateTime.now()));
> ```

Hi @FLovison,
Thanks a lot for this code. It works perfectly for filtering specific dates. Is it possible to modify this so it checks only month and year?

**Comment by akashroshan135** [comment]
Thanks for the quick reply. But ```this.year``` and ```this.month``` returns the date in utc, while I want filter with local datetime similar to ```dateLocalEquals()```. It works most of the time but if I give an input on may 1 1am, it get displayed when I filter with april instead of may

**Comment by FLovison** [comment]
I'm sorry I didn't realize...
You can compare initial part of dateLocal, it's a string in YYYY-MM-DD format ...

```dart
Expression<bool> dateYearMonthEquals(DateTime value) {
    return this.dateLocal.substring(0, 7) == value.toIso8601String().substring(0, 7);
 }
```

Alternatively you can use **_strftime_** on sqlite but it's longer ...

**Comment by akashroshan135** [comment]
I did try out a variation of that code, like this 
```dart
Expression<bool> dateYearMonthEquals(DateTime value) {
  return this.dateLocal.substring(0, 7).equals(value.toIso8601String().substring(0, 7));
}
```
Both gave me the same error
```
The method 'substring' isn't defined for the type 'Expression'.
Try correcting the name to the name of an existing method, or defining a method named 'substring'.
```
I also exprimented with other functions like contains() but most aren't working because ```dateLocal``` is of type ```Expression<String>```

**Comment by FLovison** [comment]
Oh sorry! I didn't remember that we are in a "sql where" and
before I didn't have a computer to try ... 
So, that's how it works: I use **like operator** with **%**

 ```dart
Expression<bool> dateYearMonthEquals(DateTime value) {
    return this.dateLocal.like("${value.toIso8601String().substring(0, 7)}%");
  }
```

**Comment by akashroshan135** [comment]
Wow, it worked. I just tested it out with some sample data and its filtering everything properly based on local time. Thank you so much for helping me out @FLovison

**Comment by MoeHamdan** [comment]
> > Storing the date in Iso8601 format instead of MillisecondsSinceEpoch should solve this.
> 
> That would have been a good idea, but unfortunately it'd be breaking for existing users now :(
> 
> As a starting point, you could use a custom type converter.
> 
> ```dart
> class IsoDateTimeConverter extends TypeConverter<DateTime, String> {
>   const IsoDateTimeConverter();
>   @override
>   DateTime mapToDart(String fromDb) {
>     if (fromDb == null) {
>       return null;
>     } else {
>       return DateTime.parse(fromDb);
>     }
>   }
> 
>   @override
>   String mapToSql(DateTime value) {
>     return value?.toIso8601String();
>   }
> }
> ```
> 
> In your tables, you can now do:
> 
> ```dart
> // replace this
> DateTimeColumn foo => dateTime()();
> // with this:
> TextColumn foo => text().map(const IsoDateTimeConverter())();
> ```
> 
> Arguably, moor should be able to do this out of the box. I'll think about how that could be done without breaking existing databases.

Hello
If I use this would the comparison still work like isBiggerOrEqualValue and others? I guess not right as the query will compare strings?

**Comment by North101** [comment]
@simolus3 is this worth revisiting since 2.0.0 has breaking changes?

**Comment by simolus3** [comment]
This is pretty tough. I think it might be _too_ breaking to change the default behavior (because then users would have to write a schema migration just because they've upgraded to drift 2.0). This is the worst kind of breaking change, because code breaks without you getting an analyzer warning.

Adding an option to the column builder to store it as a proper string wouldn't really work either, since `Variable(DateTime(...))` would have no way of knowing what kind of format is intended there. I think the best option now may be to move this into `driftRuntimeOptions`. So you'd do `driftRuntimeOptions.dateTimeAsStrings = true` and then all the extensions on `Expression<DateTime>` and also the builtin mappers from and to SQL would know about this.
There's still a migration involved for existing columns, but at least you'd have the freedom of doing that whenever you want to.

**Comment by North101** [comment]
Hmm yea I agree. I think a builder option would be the best way.

**Comment by North101** [comment]
> This is pretty tough. I think it might be _too_ breaking to change the default behavior (because then users would have to write a schema migration just because they've upgraded to drift 2.0). This is the worst kind of breaking change, because code breaks without you getting an analyzer warning.
> 
> Adding an option to the column builder to store it as a proper string wouldn't really work either, since `Variable(DateTime(...))` would have no way of knowing what kind of format is intended there. I think the best option now may be to move this into `driftRuntimeOptions`. So you'd do `driftRuntimeOptions.dateTimeAsStrings = true` and then all the extensions on `Expression<DateTime>` and also the builtin mappers from and to SQL would know about this. There's still a migration involved for existing columns, but at least you'd have the freedom of doing that whenever you want to.

Would something like this work:
```dart
@DriftDatabase()
class Database extends _$Database {
  @override
  final typeSystem = const SqlTypeSystem.withDefaultsUtc();
}
```

**Comment by simolus3** [comment]
I like the idea of adding an override. Since the option at runtime needs to be consistent with the option at compile-time (so that `DATETIME` columns get interpreted as text in the query analyzer), this override could ideally be generated by `drift_dev` in `_$Database`.

**Comment by simolus3** [comment]
Finally, support for storing datetimes as text has landed in `develop` :tada:  Thank you for all your suggestions (and patience :D).

As always, I'd love to get feedback on this if anyone wants to try out already. The documentation gives an overview of how to migrate between the two modes [here](https://62deea1276648b05c27d0cbe--moor.netlify.app/docs/getting-started/advanced_dart_tables/#datetime-options).

This feature touches a lot of parts in drift and while I have added symmetric test for all existing datetime tests, it is still possible that I have missed something. So, if you run into any issues with this (or if the documentation isn't clear to you), please let me hear about it!

**Comment by North101** [comment]
> Finally, support for storing datetimes as text has landed in `develop` 🎉 Thank you for all your suggestions (and patience :D).
> 
> As always, I'd love to get feedback on this if anyone wants to try out already. The documentation gives an overview of how to migrate between the two modes [here](https://62deea1276648b05c27d0cbe--moor.netlify.app/docs/getting-started/advanced_dart_tables/#datetime-options).
> 
> This feature touches a lot of parts in drift and while I have added symmetric test for all existing datetime tests, it is still possible that I have missed something. So, if you run into any issues with this (or if the documentation isn't clear to you), please let me hear about it!

So i have changed my build.yaml to:
```
targets:
  $default:
    builders:
      drift_dev:
        options:
          apply_converters_on_variables: true
          generate_values_in_copy_with: true
          named_parameters: true
          override_hash_and_equals_in_result_sets: true
          generate_connect_constructor: true
          store_date_time_values_as_text: true
```

Then called `flutter pub run build_runner build --delete-conflicting-outputs`.

But when I create a new database it seems to be creating the columns as `INTEGER`. I'm using `.drift` files if that could be the cause.

**Comment by North101** [comment]
Or maybe `validateDatabaseSchema` doesn't recognise the option

Edit: I think this is the issue. Everything seems to be working fine

**Comment by North101** [comment]
I was getting:
```
Unhandled Exception: Invalid argument(s): Invalid call to read(): Instance of '_LazyExpression<String>'. This result set does not have a column for that expression.
```

Because I had a query like this:
```dart
query = selectOnly(table);
query.addColumns([datetimeColumn.date]);
return query.get().map((e) => e.read(datetimeColumn.date));
```

I fixed it by changing it to:
```dart
final dateColumn = datetimeColumn.date;
query = selectOnly(table);
query.addColumns([dateColumn]);
return query.get().map((e) => e.read(dateColumn));
```

**Comment by North101** [comment]
```dart
class _DefaultValueSerializer extends ValueSerializer {
  const _DefaultValueSerializer();

  @override
  T fromJson<T>(dynamic json) {
    if (json == null) {
      return null as T;
    }

    final _typeList = <T>[];

    if (_typeList is List<DateTime?>) {
      return DateTime.fromMillisecondsSinceEpoch(json as int) as T;
    }
```

Should the `fromJson` method also handle datetime Strings?

**Comment by simolus3** [comment]
Thanks a lot for trying this out!

> `Unhandled Exception: Invalid argument(s): Invalid call to read(): Instance of '_LazyExpression<String>'. This result set does not have a column for that expression.`

Good catch, fixed in 77e7b99f2f636d9f3b0ca1db62216ec4cc0b0807.

> Or maybe `validateDatabaseSchema` doesn't recognise the option

Nice catch, fixed in 9158c312405a676dbfabdb219afae6d871793b63. A similar problem arises for exported schemas and generated tests, I have fixed that as well in 0cb352768079594bd18d062eb41e3dcd9ec92893.

> Should the `fromJson` method also handle datetime Strings?

`fromJson` can support both of course, but the behavior in `toJson` is probably more interesting. At the moment, the JSON serializer is not part of the database's options. I'll take a look at how tricky it will be to change that :D

**Comment by zhangruiyu** [comment]
class IsoDateTimeConverter extends TypeConverter<DateTime, int> {
  const IsoDateTimeConverter();

  @override
  DateTime fromSql(int fromDb) {
    return DateTime.fromMicrosecondsSinceEpoch(fromDb);
  }

  @override
  int toSql(DateTime value) {
    return value.microsecondsSinceEpoch;
  }
}
IntColumn get createTime => integer().map(IsoDateTimeConverter())();
